core/lib/instrument.c

```diff
@@ -1955,6 +1955,14 @@ instrument_module_load_trigger(app_pc pc)
         if (ma != NULL && !TEST(MODULE_LOAD_EVENT, ma->flags)) {
             /* switch to write lock */
             os_get_module_info_unlock();
+
+            /* i#3385: re-initialize dynamic information, because it failed
+             * during the first flat-mmap that loaded the module.
+             */
+            if (!ma->os_data.have_dynamic_info) {
+                os_module_update_dynamic_info(ma->start, ma->end - ma->start, false);
```

Comment:
OK, this seems reasonable: at first execution if we don't have the data we try again.  And we'll get this far because the faults on the first instance are all inside try..excepts?

Should we leave the assert curiosities in place?  Probably.

And the SIGBUS you reported is from inside the debugger and it's caught by the try..except and is not fatal, right?

This code is currently wrapped in CLIENTS_EXIST(): do we need this when there are no clients?  Or we assume DR itself doesn't care about this module data?  I assume the latter: maybe worth adding a comment that we deliberately do not bother to re-try if there are no clients.

Another request is to go add a comment by that fault point in that try..except and point to i#3385 and explain that we'll try again in instrument_module_load_trigger() at first execution.

Reply:
> OK, this seems reasonable: at first execution if we don't have the data we try again. And we'll get this far because the faults on the first instance are all inside try..excepts?

Yes, module parsing is implemented inside try..excepts, so all bugs with unexpected/alternatively formed executables won't crash DR.

> And the SIGBUS you reported is from inside the debugger and it's caught by the try..except and is not fatal, right?

It is debug output for failed test string readablity in parsing header wrapped in try..except. 

> This code is currently wrapped in CLIENTS_EXIST(): do we need this when there are no clients?

I don't know whether DR uses os_nodule_data_t itself. When I first encountered this error, first of all I tried to launch DR with dummy/empty-clients/without-client, and it didn't crashed. So I guess it can be left only for client-usecases.

Reply:
Updated comments